docs/architecture-overview.rst

```diff
@@ -0,0 +1,311 @@
+translations.
+
+In this overview we'll mainly discuss the core Zulip server and
+web application.
+
+
+Usage assumptions and concepts
+==============================
+
+Zulip is a real-time web-based chat application meant for companies
+and similar groups ranging in size from a small team to more than a
+thousand users. It features real-time notifications, message
+persistence and search, public group conversations (*streams*),
+invite-only streams, private one-on-one and group conversations,
+inline image previews, team presence/a buddy list, a rich API, and
+several integrations with other services. The maintainer team aims to
+support users who connect to Zulip using dedicated iOS, Android,
+Linux, Windows, and Mac OS X clients, as well as people using modern
+web browsers or dedicated Zulip API clients.
+
+A server can host multiple Zulip *realms* (organizations) at the same
+domain, each of which is a private chamber with its own users,
+streams, customizations, and so on. This means that one person might
+be a user of multiple Zulip realms. The administrators of a realm can
+choose whether to allow anyone to register an account and join, or
+only allow people who have been invited, or restrict registrations to
+members of particular groups (using email domain names or corporate
+single-sign-on login for verification). For more on scalability and
+security considerations, see `Zulip in production
+<https://github.com/zulip/zulip/blob/master/README.prod.md>`__.
+
+The default Zulip home screen is like a chronologically ordered inbox;
+it displays messages, starting at the oldest message that the user
+hasn't viewed yet. The home screen displays the most recent messages
+in all the streams a user has joined (except for the streams they've
+muted), as well as private messages from other users, in strict
+chronological order. A user can "narrow" to view only the messages in
+a single stream, and can further narrow to focus on a *topic* (thread)
+within that stream. Each narrow has its own URL.
+
+Zulip's philosophy is to provide sensible defaults but give the user
+fine-grained control over their incoming information flow; a user can
+mute topics and streams, and can make fine-grained choices to reduce
+real-time notifications they find irrelevant.
+
+
+
+Components
+==========
+
+
+Tornado and Django
+------------------
+
+We use both the `Tornado <http://www.tornadoweb.org>`__ and `Django
+<https://www.djangoproject.com/>`__ Python web frameworks.
+
+Django is the main web application server; Tornado runs the
+server-to-client real-time push system. The app servers are configured
+by the Supervisor configuration (which explains how to start the
+server processes; see "Supervisor" below) and the nginx configuration
+(which explains which HTTP requests get sent to which app server).
+
+Tornado is an asynchronous server, so it can scale really well, but
+asynchronous programming has a learning curve you might need to watch
```

Comment:
I'd ignore the learning curve part entirely.

Tornado has a really specific use: holding open tens of thousands of long-lived (long-polling or websocket) connections. For this reason, it's responsible for message delivery, but not much else. We try to avoid any blocking calls in Tornado because we don't want to delay delivery to thousands of other connections (as this would make Zulip very much not real-time).